Sources/edge/domain/models/DeviceModels.swift

```diff
@@ -0,0 +1,243 @@
+        allDevices.append(contentsOf: ethernet)
+        self.devices = allDevices
+    }
+
+    func toJSON() throws -> String {
+        // Since Device is a protocol, we need to handle heterogeneous collection
+        // One approach is to create a dictionary with device types as keys
+        var devicesByType: [String: [Any]] = [:]
+
+        for device in devices {
+            if let usbDevice = device as? USBDevice {
+                if devicesByType["usbDevices"] == nil {
+                    devicesByType["usbDevices"] = []
+                }
+                devicesByType["usbDevices"]?.append(usbDevice)
+            } else if let ethernetDevice = device as? EthernetInterface {
+                if devicesByType["ethernetDevices"] == nil {
+                    devicesByType["ethernetDevices"] = []
+                }
+                devicesByType["ethernetDevices"]?.append(ethernetDevice)
+            }
+            // Add more device types as needed
+        }
+
+        // If there are no devices, return an empty object without pretty printing
+        if devicesByType.isEmpty {
+            return "{}"
+        }
+
+        let encoder = JSONEncoder()
+        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
+
+        // We need to convert to a custom structure for encoding
+        let encodableDict = EncodableDict(dict: devicesByType)
+        let data = try encoder.encode(encodableDict)
+        return String(data: data, encoding: .utf8) ?? "{}"
+    }
+
+    func toHumanReadableString() -> String {
+        var result = ""
+
+        // Group devices by type
+        let usbDevices = devices.compactMap { $0 as? USBDevice }
+        let ethernetDevices = devices.compactMap { $0 as? EthernetInterface }
+
+        // Add USB devices section
+        if !usbDevices.isEmpty {
+            result += "\nUSB Devices:"
+            for device in usbDevices {
+                result += "\n" + device.toHumanReadableString()
+            }
+        }
+
+        // Add Ethernet devices section
+        if !ethernetDevices.isEmpty {
+            if !result.isEmpty {
+                result += "\n"
+            }
+            result += "\nEthernet Interfaces:"
+            for device in ethernetDevices {
+                result += "\n" + device.toHumanReadableString()
+            }
+        }
+
+        return result.isEmpty ? "No devices found." : result
+    }
+
+    // Helper struct for encoding heterogeneous collections
+    private struct EncodableDict: Encodable {
+        enum AnyDevice: Encodable {
+          case usb(USBDevice)
+          case ethernet(EthernetDevice)
+          
+          func encode(to encoder: Encoder) throws {
+              switch self {
+              case .usb(let device):
+                try device.encode(to: encoder)
+              case .ethernet(let device):
+                try device.encode(to: encoder)
+              }
+          }
+        }
+        
+        let dict: [String: [AnyDevice]]
```

Comment:
We could now also:
```
typealias EncodableDict = [String: [AnyDevice]]
```

Reply:
I updated the implementation and I tried this but it didn't quite work. needs more massaging. going to resolve this for now.